packages/account/src/BiconomySmartAccountV2.ts

```diff
@@ -154,33 +238,277 @@ export class BiconomySmartAccountV2 extends BaseSmartAccount {
+      )
+    }
+
+    const spender = tokenPaymasterRequest?.spender
+    Logger.log('fee token approval to be checked and added for spender: ', spender)
+
+    if (!spender || spender == ethers.constants.AddressZero) {
+      throw new Error(
+        'Invalid or missing spender address. Sepnder address must be part of tokenPaymasterRequest'
+      )
+    }
+  }
+
+  /**
+   *
+   * @param userOp partial user operation without signature and paymasterAndData
+   * @param tokenPaymasterRequest This dto provides information about fee quote. Fee quote is received from earlier request getFeeQuotesOrData() to the Biconomy paymaster.
+   *  maxFee and token decimals from the quote, along with the spender is required to append approval transaction.
+   * @notice This method should be called when gas is paid in ERC20 token using TokenPaymaster
+   * @description Optional method to update the userOp.calldata with batched transaction which approves the paymaster spender with necessary amount(if required)
+   * @returns updated userOp with new callData, callGasLimit
+   */
+  async buildTokenPaymasterUserOp(
+    userOp: Partial<UserOperation>,
+    tokenPaymasterRequest: BiconomyTokenPaymasterRequest
+  ): Promise<Partial<UserOperation>> {
+    this.validateUserOpAndRequest(userOp, tokenPaymasterRequest)
+    try {
+      let batchTo: Array<string> = []
+      let batchValue: Array<string | BigNumberish> = []
+      let batchData: Array<string> = []
+
+      let newCallData = userOp.callData
+      Logger.log('received information about fee token address and quote ', tokenPaymasterRequest)
+
+      if (this.paymaster && this.paymaster instanceof BiconomyPaymaster) {
+        // Make a call to paymaster.buildTokenApprovalTransaction() with necessary details
+
+        // Review: might request this form of an array of Transaction
+        const approvalRequest: Transaction = await (
+          this.paymaster as IHybridPaymaster<SponsorUserOperationDto>
+        ).buildTokenApprovalTransaction(tokenPaymasterRequest, this.provider)
+        Logger.log('approvalRequest is for erc20 token ', approvalRequest.to)
+
+        if (approvalRequest.data == '0x' || approvalRequest.to == ethers.constants.AddressZero) {
+          return userOp
+        }
+
+        if (!userOp.callData) {
+          throw new Error('Userop callData cannot be undefined')
+        }
+
+        const account = await this._getAccountContract()
+
+        const decodedDataSmartWallet = account.interface.parseTransaction({
+          data: userOp.callData.toString()
+        })
+        if (!decodedDataSmartWallet) {
+          throw new Error('Could not parse call data of smart wallet for userOp')
+        }
+
+        const smartWalletExecFunctionName = decodedDataSmartWallet.name
+
+        if (smartWalletExecFunctionName === 'executeCall') {
+          Logger.log('originally an executeCall for Biconomy Account')
```

Comment:
Update these logs accordingly. Or better use

Logger.log(\`Originally an ${smartAccountExecFunctionName} method call for Biconomy Account\`)

This can be put before if statement as common log for both if and else block.

Reply:
done